curvefs/src/metaserver/dentry_storage.cpp

```diff
@@ -287,50 +542,145 @@ MetaStatusCode DentryStorage::Insert(const DentryVec& vec, bool merge) {
 
     DentryVector vector(&oldVec);
     vector.Merge(vec);
-    s = kvStorage_->SSet(table4Dentry_, skey, oldVec);
+    s = txn->SSet(table4Dentry_, skey, oldVec);
+    if (!s.ok()) {
+        if (!txn->Rollback().ok()) {
+            LOG(ERROR) << "Rollback insert dentry transaction failed, status = "
+                       << s.ToString();
+        }
+        LOG(ERROR) << "Insert dentry vector to tranasction failed, status = "
+                   << s.ToString();
+        return MetaStatusCode::STORAGE_INTERNAL_ERROR;
+    }
+    s = SetAppliedIndex(txn.get(), logIndex);
+    if (!s.ok()) {
+        if (!txn->Rollback().ok()) {
+            LOG(ERROR) << "Rollback insert dentry transaction failed, status = "
+                       << s.ToString();
+        }
+        LOG(ERROR) << "Insert applied index to tranasction failed, status = "
+                   << s.ToString();
+        return MetaStatusCode::STORAGE_INTERNAL_ERROR;
+    }
+    std::uint64_t count = nDentry_;
+    vector.Confirm(&count);
+    s = SetDentryCount(txn.get(), count);
+    if (!s.ok()) {
+        if (!txn->Rollback().ok()) {
+            LOG(ERROR) << "Rollback insert dentry transaction failed, status = "
+                       << s.ToString();
+        }
+        LOG(ERROR) << "Insert dentry count to transaction failed, status = "
+                   << s.ToString();
+        return MetaStatusCode::STORAGE_INTERNAL_ERROR;
+    }
+    s = txn->Commit();
     if (!s.ok()) {
+        if (!txn->Rollback().ok()) {
+            LOG(ERROR) << "Rollback insert dentry transaction failed, status = "
+                       << s.ToString();
+        }
```

Comment:
consider refactor these code with `do { } while (false)`, like I suggested in inode_storage.cpp